test/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -40,82 +40,126 @@ class SingleSignOnProfileVerifier(val response: Node) {
+                    ?.textContent
+
+            if (audience != SP_ISSUER)
+                throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLProfiles_4_1_4_2_k,
+                        "Audience",
+                        audience,
+                        SP_ISSUER)
+        }
+    }
+
+    private fun verifySingleLogoutIncludesSessionIndex(it: Node) {
+        if (idpMetadata.descriptor == null) return
+
+        if (!idpMetadata.descriptor!!.singleLogoutServices.isNotEmpty()) return
+
+        it.children("AuthnStatement").forEach {
+            if (it.attributes.getNamedItem("SessionIndex") == null) {
+                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_j,
+                        message = "Single Logout support found in IdP metadata, but no SessionIndex was" +
+                                " found.")
             }
         }
     }
-}
\ No newline at end of file
+
+    private fun verifyAssertionsHaveAuthnStmts(it: Node) {
+        if (it.children("AuthnStatement").isEmpty()) {
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_i,
+                    message = "A bearer Assertion was found without an AuthnStatement.")
+        }
+    }
+
+    @Suppress("ComplexCondition")
+    private fun verifyBearerSubjectConfirmations(it: Node) {
+        val bearerSubjectConfirmations = it.children("Subject")[0].children("SubjectConfirmation")
+                .filter {
+                    it.attributes.getNamedItem("Method").textContent ==
+                            "urn:oasis:names:tc:SAML:2.0:cm:bearer"
+                }.toList()
+
+        //todo - We can't throw an exception here because the spec says there could be assertions without bearer
+        // SubjectConfirmation
+        if (bearerSubjectConfirmations.isEmpty())
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_g,
+                    message = "No bearer SubjectConfirmation elements were found.")
+
+        // Check if there is one SubjectConfirmationData with a Recipient, InResponseTo and NotOnOrAfter
+        if (bearerSubjectConfirmations
+                        .filter { it.children("SubjectConfirmationData").isNotEmpty() }
+                        .flatMap { it.children("SubjectConfirmationData") }
+                        .none {
+                            it.attributes.getNamedItem("Recipient").textContent == ACS_URL &&
+                                    it.attributes.getNamedItem("InResponseTo").textContent == ID &&
+                                    it.attributes.getNamedItem("NotOnOrAfter") != null &&
+                                    it.attributes.getNamedItem("NotBefore") == null
+                        }) {
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h,
+                    message = "There were no bearer SubjectConfirmation elements that matched the criteria below.")
+        }
+    }
+
+    private fun verifyHasSubject(it: Node) {
```

Comment:
✏️ Do you think we could change the parameter on these private functions to be more descriptive now that we can? Like `subjectNode` in this instance.

Reply:
will do.